pulsar/pulsar.sh

```diff
@@ -0,0 +1,230 @@
+readonly PULSAR_HOME="/opt/apache-pulsar-${PULSAR_VERSION}"
+
+readonly WEB_SERVICE_URL_PORT="$(/usr/share/google/get_metadata_value attributes/web-service-url-port || echo 8080)"               #default value, not recommended to change
+readonly WEB_SERVICE_URL_TLS_PORT="$(/usr/share/google/get_metadata_value attributes/web-service-url-tls-port || echo 8443)"       #default value, not recommended to change
+readonly BROKER_SERVICE_URL_PORT="$(/usr/share/google/get_metadata_value attributes/broker-service-url-port || echo 6650)"         #default value, not recommended to change
+readonly BROKER_SERVICE_URL_TLS_PORT="$(/usr/share/google/get_metadata_value attributes/broker-service-url-tls-port || echo 6651)" #default value, not recommended to change
+
+ZOOKEEPER_ADDRESSES=""
+ZOOKEEPER_ADDRESS=""
+
+WEB_SERVICE_URL=""
+WEB_SERVICE_URL_TLS=""
+BROKER_SERVICE_URL=""
+BROKER_SERVICE_URL_TLS=""
+
+function retry_command() {
+  local retry_backoff=(1 1 2 3 5 8 13 21 34 55 89 144)
+  local -a cmd=("$@")
+  echo "About to run '${cmd[*]}' with retries..."
+
+  local installation_succeeded=0
+  for ((i = 0; i < ${#retry_backoff[@]}; i++)); do
+    if eval "${cmd[@]}"; then
+      installation_succeeded=1
+      break
+    else
+      local sleep_time=${retry_backoff[$i]}
+      echo "'${cmd[*]}' attempt $((i + 1)) failed! Sleeping ${sleep_time}." >&2
+      sleep "${sleep_time}"
+    fi
+  done
+
+  if ! ((installation_succeeded)); then
+    echo "Final attempt of '${cmd[*]}'..."
+    # Let any final error propagate all the way out to any error traps.
+    "${cmd[@]}"
+  fi
+}
+
+function edit() {
+  if [ "$#" == "3" ]; then
+    grep -q "$1" "$3" && sed -i "s/"$1"/"$2"/" "$3" || echo "$2" >>"$3"
+  else
+    echo "$1" >>"$2"
+  fi
+}
+
+function err() {
+  echo "[$(date +'%Y-%m-%dT%H:%M:%S%z')]: $@" >&2
+  return 1
+}
+
+function retry_apt_command() {
+  cmd="$1"
+  for ((i = 0; i < 10; i++)); do
+    if eval "$cmd"; then
+      return 0
+    fi
+    sleep 5
+  done
+  return 1
+}
+
+function update_apt_get() {
+  retry_apt_command "apt-get update"
+}
+
+function fetch_zookeeper_list() {
+  local zookeeper_client_port
+  zookeeper_client_port=$(grep 'clientPort' /etc/zookeeper/conf/zoo.cfg |
+    tail -n 1 |
+    cut -d '=' -f 2)
+
+  local zookeeper_list
+  zookeeper_list=$(grep '^server\.' /etc/zookeeper/conf/zoo.cfg |
+    cut -d '=' -f 2 |
+    cut -d ':' -f 1 |
+    sort |
+    uniq |
+    sed "s/$/:${zookeeper_client_port}/" |
+    xargs echo |
+    sed "s/ /,/g")
+
+  ZOOKEEPER_ADDRESSES="${zookeeper_list}"
+
+  # If attempt fails, error out.
+  if [[ -z "${zookeeper_list}" ]]; then
+    err 'Failed to find configured Zookeeper list. Please remember to include Zookeeper optional component'
+  fi
+
+  ZK_ADDRS="${zookeeper_list}"
+  ZK_ADDRS=${ZK_ADDRS//,/$'\n'}
+  ZK_ADDRS=($ZK_ADDRS)
+
+  for i in "${!ZK_ADDRS[@]}"; do
+    local serverNo=$i
+    serverNo=$((serverNo + 1))
+    edit "server.${serverNo}=${ZK_ADDRS[$i]}" conf/zookeeper.conf
```

Comment:
In my test, it failed with `conf/zookeeper.conf: No such file or directory`. Also, the name of this function is `fetch_zookeeper_list`, which feels like read-only, but actually it is editing config. So, consider rename it to `configure_zookeeper`.